src/gui/plugins/axes/AxesConfig.cc

```diff
@@ -0,0 +1,285 @@
+    // rendering calls here
+    UpdateOriginArrows();
+  }
+
+  // Standard event processing
+  return QObject::eventFilter(_obj, _event);
+}
+
+void AxesConfig::UpdateActiveAxes()
+{
+  for(auto vis : this->dataPtr->activeAxesMap)
+  {
+    auto visAxes = std::dynamic_pointer_cast<rendering::AxisVisual>(
+          this->dataPtr->scene->VisualByName(vis.first + "Axes"));
+
+    auto visEntity = std::dynamic_pointer_cast<rendering::Visual>(
+          this->dataPtr->scene->VisualByName(vis.first));
+
+    if (visEntity && visAxes){
+      math::Pose3d pose_entity = visEntity->LocalPose();
+      visAxes->SetLocalPose(vis.second + pose_entity);
+    }
+  }
+}
+
+/////////////////////////////////////////////////
+void AxesConfig::UpdateOriginArrows()
+{
+  // Load axes if they don't already exist
+  this->LoadAxesbyName(this->dataPtr->name_axes);
+
+  // If axes were not loaded successfully, don't update
+  if (!this->dataPtr->axes)
+    return;
+
+  // Save the axesVisual in the structure if it doesn't exist or update the pose
+  auto it = this->dataPtr->activeAxesMap.find(this->dataPtr->name_axes);
+  if (it == this->dataPtr->activeAxesMap.end())
+  {
+    this->dataPtr->activeAxesMap.insert(
+      std::pair<std::string, math::Pose3d>(this->dataPtr->name_axes, this->dataPtr->pose));
+  } else {
+    it->second = this->dataPtr->pose;
+  }
+
+  // Updating the poses
+  this->UpdateActiveAxes();
+
+  // update visibility
+  this->dataPtr->axes->SetVisible(this->dataPtr->visible);
+  // Update type: arrow or line
+  if (this->dataPtr->visible)
+  {
+    this->dataPtr->axes->ShowAxisHead(this->dataPtr->isArrow);
+  }
+  // update scale
+  this->dataPtr->axes->SetLocalScale(1, 1, this->dataPtr->length * 2);
+}
+
+const QStringList AxesConfig::comboList()
+{
+    return itemComboList;
+}
+
+void AxesConfig::SetComboList(const QStringList &comboList)
+{
+   if (itemComboList != comboList)
+   {
+       itemComboList = comboList;
+       emit ComboListChanged();
+   }
+}
+
+void AxesConfig::LoadAxesbyName(const std::string & name)
+{
+  if (this->dataPtr->name_axes.compare(name) == 0 && this->dataPtr->axes)
+    return;
+
+  this->dataPtr->name_axes = name;
+  if (this->dataPtr->name_axes.compare(worldOriginAxes) != 0)
+  {
+    this->dataPtr->axes = std::dynamic_pointer_cast<rendering::AxisVisual>(
+        this->dataPtr->scene->VisualByName(this->dataPtr->name_axes + "Axes"));
+  } else {
+    this->dataPtr->axes = std::dynamic_pointer_cast<rendering::AxisVisual>(
+        this->dataPtr->scene->VisualByName(worldOriginAxes));
+  }
+
+  if (!this->dataPtr->axes)
+  {
+    rendering::VisualPtr root = this->dataPtr->scene->RootVisual();
+
+    auto axes = this->dataPtr->scene->CreateAxisVisual(this->dataPtr->name_axes + "Axes");
+    root->AddChild(axes);
+  }
+}
+
+void AxesConfig::onCurrentIndexChanged(int _index)
+{
+  std::string name_entity = itemComboList[_index].toStdString();
+  LoadAxesbyName(name_entity);
```

Comment:
This needs to be done in the rendering thread, otherwise we get occasional crashes like this one I just got:

```
Invalid parameter at line 374 (@foreach). 'hlms_lights_directional' is not a number nor a variable
terminate called after throwing an instance of 'Ogre::RenderingAPIException'
  what():  OGRE EXCEPTION(3:RenderingAPIException): Fragment Program 537100288PixelShader_ps failed to compile. See compile log above for details. in GLSLShader::compile at /var/lib/jenkins/workspace/ogre-2.1-debbuilder/repo/RenderSystems/GL3Plus/src/GLSL/OgreGLSLShader.cpp (line 308)
```

To use the rendering thread, listen to events like this:

https://github.com/ignitionrobotics/ign-gazebo/blob/aea1accff6d8daee0358ae9025e2a9797d292844/src/gui/plugins/transform_control/TransformControl.cc#L94-L95

And look for the `Render` event:

https://github.com/ignitionrobotics/ign-gazebo/blob/aea1accff6d8daee0358ae9025e2a9797d292844/src/gui/plugins/transform_control/TransformControl.cc#L220-L222